test-utils/runtime/src/substrate_test_pallet.rs

```diff
@@ -0,0 +1,292 @@
+				None => storage::unhashed::kill(&key),
+			}
+			Ok(())
+		}
+
+		#[pallet::call_index(4)]
+		#[pallet::weight(100)]
+		pub fn storage_change(
+			origin: OriginFor<T>,
+			key: Vec<u8>,
+			value: Option<Vec<u8>>,
+		) -> DispatchResult {
+			frame_system::ensure_signed(origin)?;
+			match value {
+				Some(value) => storage::unhashed::put_raw(&key, &value),
+				None => storage::unhashed::kill(&key),
+			}
+			Ok(())
+		}
+
+		#[pallet::call_index(5)]
+		#[pallet::weight(100)]
+		pub fn offchain_index_set(
+			origin: OriginFor<T>,
+			key: Vec<u8>,
+			value: Vec<u8>,
+		) -> DispatchResult {
+			frame_system::ensure_signed(origin)?;
+			sp_io::offchain_index::set(&key, &value);
+			Ok(())
+		}
+
+		#[pallet::call_index(6)]
+		#[pallet::weight(100)]
+		pub fn offchain_index_clear(origin: OriginFor<T>, key: Vec<u8>) -> DispatchResult {
+			frame_system::ensure_signed(origin)?;
+			sp_io::offchain_index::clear(&key);
+			Ok(())
+		}
+
+		#[pallet::call_index(7)]
+		#[pallet::weight(100)]
+		pub fn store(origin: OriginFor<T>, data: Vec<u8>) -> DispatchResult {
+			frame_system::ensure_signed(origin)?;
+			let content_hash = sp_io::hashing::blake2_256(&data);
+			let extrinsic_index: u32 =
+				storage::unhashed::get(well_known_keys::EXTRINSIC_INDEX).unwrap();
+			sp_io::transaction_index::index(extrinsic_index, data.len() as u32, content_hash);
+			Ok(())
+		}
+
+		#[pallet::call_index(8)]
+		#[pallet::weight(100)]
+		pub fn deposit_log_digest_item(
+			_origin: OriginFor<T>,
+			log: sp_runtime::generic::DigestItem,
+		) -> DispatchResult {
+			<frame_system::Pallet<T>>::deposit_log(log);
+			Ok(())
+		}
+	}
+
+	#[pallet::validate_unsigned]
+	impl<T: Config> ValidateUnsigned for Pallet<T> {
+		type Call = Call<T>;
+
+		fn validate_unsigned(_source: TransactionSource, call: &Self::Call) -> TransactionValidity {
+			log::trace!(target: LOG_TARGET, "validate_unsigned {call:?}");
+			validate_runtime_call(call)
+		}
+	}
+}
+
+pub fn balance_of_key(who: AccountId) -> Vec<u8> {
+	who.to_keyed_vec(BALANCE_OF)
+}
+
+pub fn balance_of(who: AccountId) -> u64 {
+	storage::hashed::get_or(&blake2_256, &balance_of_key(who), 0)
+}
+
+pub fn authorities() -> Vec<AuthorityId> {
+	<Authorities<Runtime>>::get()
+}
+
+pub fn get_block_number() -> Option<BlockNumber> {
+	<Number<Runtime>>::get()
+}
+
+use codec::Encode;
+use sp_runtime::transaction_validity::{
+	InvalidTransaction, TransactionSource, TransactionValidity, ValidTransaction,
+};
+
+pub fn validate_runtime_call<T: pallet::Config>(call: &pallet::Call<T>) -> TransactionValidity {
+	log::trace!(target: LOG_TARGET, "validate_runtime_call {call:?}");
+	match call {
+		Call::transfer { transfer, signature, exhaust_resources_when_not_first } => {
+			let extrinsic_index: u32 =
+				storage::unhashed::get(well_known_keys::EXTRINSIC_INDEX).unwrap_or_default();
+
+			if *exhaust_resources_when_not_first && extrinsic_index != 0 {
+				return InvalidTransaction::ExhaustsResources.into()
+			}
```

Comment:
Generally we should not require this "when first extrinsic" part anymore. We should just rewrite tests to push the exhaust extrinsic only when they need it.

Reply:
Done